src/core_functions/scalar/list/list_reduce.cpp

```diff
@@ -0,0 +1,241 @@
+#include "duckdb/core_functions/scalar/list_functions.hpp"
+#include "duckdb/core_functions/lambda_functions.hpp"
+#include "duckdb/planner/expression/bound_cast_expression.hpp"
+#include "duckdb/planner/expression/bound_function_expression.hpp"
+
+namespace duckdb {
+
+struct ReduceExecuteInfo {
+	ReduceExecuteInfo(LambdaFunctions::LambdaInfo &info, ClientContext &context) : left_slice(*info.child_vector) {
+		SelectionVector left_vector(info.row_count);
+		active_rows.Resize(0, info.row_count);
+		active_rows.SetAllValid(info.row_count);
+
+		right_sel.Initialize(info.row_count);
+		left_sel.Initialize(info.row_count);
+		active_rows_sel.Initialize(info.row_count);
+
+		idx_t old_count = 0;
+		idx_t new_count = 0;
+
+		// Initialize the left vector from list entries and the active rows
+		auto data = active_rows.GetData();
+
+		for (idx_t entry_idx = 0; old_count < info.row_count; entry_idx++) {
+			if (data[entry_idx] == 0) {
+				old_count += 64;
+				continue;
+			}
+
+			for (idx_t j = 0; j < 64 && entry_idx + j < info.row_count; j++) {
```

Comment:
I think that we can remove the validity mask logic here and loop over `row_count` instead, as we initialise `active_rows` to `SetAllValid`.